dpnp/backend/extensions/lapack/heevd_batch.cpp

```diff
@@ -0,0 +1,207 @@
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include "heevd_batch.hpp"
+#include "common_helpers.hpp"
+
+namespace dpnp::extensions::lapack
+{
+namespace mkl_lapack = oneapi::mkl::lapack;
+namespace type_utils = dpctl::tensor::type_utils;
+
+template <typename T, typename RealT>
+static sycl::event heevd_batch_impl(sycl::queue &exec_q,
+                                    const oneapi::mkl::job jobz,
+                                    const oneapi::mkl::uplo upper_lower,
+                                    const std::int64_t batch_size,
+                                    const std::int64_t n,
+                                    char *in_a,
+                                    char *out_w,
+                                    const std::vector<sycl::event> &depends)
+{
+    type_utils::validate_type_for_device<T>(exec_q);
+    type_utils::validate_type_for_device<RealT>(exec_q);
+
+    T *a = reinterpret_cast<T *>(in_a);
+    RealT *w = reinterpret_cast<RealT *>(out_w);
+
+    std::int64_t a_size = n * n;
+    std::int64_t w_size = n;
+
+    const std::int64_t lda = std::max<size_t>(1UL, n);
+
+    // Get the number of independent linear streams
+    std::int64_t n_linear_streams =
+        (batch_size > 16) ? 4 : ((batch_size > 4 ? 2 : 1));
+
+    const std::int64_t scratchpad_size =
+        mkl_lapack::heevd_scratchpad_size<T>(exec_q, jobz, upper_lower, n, lda);
+    T *scratchpad = nullptr;
+
+    // Get padding size to ensure memory allocations are aligned to 256 bytes
+    // for better performance
+    std::int64_t padding = 256 / sizeof(T);
+
+    if (scratchpad_size > 0) {
+        // Calculate the total scratchpad memory size needed for all linear
+        // streams with proper alignment
+        size_t alloc_scratch_size =
+            helper::round_up_mult(n_linear_streams * scratchpad_size, padding);
+
+        // Allocate memory for the total scratchpad
+        scratchpad = sycl::malloc_device<T>(alloc_scratch_size, exec_q);
+        if (!scratchpad)
+            throw std::runtime_error("Device allocation for scratchpad failed");
```

Comment:
Do we need to have similar check in `heevd_impl` if memory allocation for scratchpad returns `nullptr`?